File: src/04_MedianOfTwoSortedArrays.java
class Solution {
    
    // O(m+n) solution using merge from mergesort
    private int[] merge(int[] nums1, int[] nums2){
        int[] merged = new int[nums1.length + nums2.length];
        
        int i = 0, j = 0, k = 0;
        
        while (i < nums1.length && j < nums2.length){
            if (nums1[i] < nums2[j]){
                merged[k++] = nums1[i++];
            } else {
                merged[k++] = nums2[j++];
            }
        }
        
        while (i < nums1.length){
            merged[k++] = nums1[i++];
        }
        
        while (j < nums2.length){
            merged[k++] = nums2[j++];
        }
        
        return merged;
    }
    
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        int[] merged = merge(nums1, nums2);
        int n = merged.length;
        if (n == 0) return 0.0;
        if (n % 2 == 0){
            return (merged[n/2] + merged[n/2-1])/2.0;
        } else {
            return merged[n/2];
        }
    }
}

File: src/04-median-of-two-sorted-arrays.go
/*
O(n+m) solution. Should be O(log(n+m)). Needs improvement https://www.geeksforgeeks.org/median-two-sorted-arrays-different-sizes-ologminn-m/
*/

func getMedian(nums []int) float64 {
    size := len(nums)
    if size % 2 == 0 {
        return float64(nums[size/2-1] + nums[size/2]) / 2
    }
    return float64(nums[size/2])
}

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
    ptr1, ptr2, index := 0, 0, 0 
    size := len(nums1) + len(nums2)
    merged := make([]int, size)
    for ptr1 < len(nums1) && ptr2 < len(nums2) {
        if nums1[ptr1] < nums2[ptr2] {
            merged[index] = nums1[ptr1]
            ptr1 = ptr1 + 1
        } else {
            merged[index] = nums2[ptr2]
            ptr2 = ptr2 + 1
        }
        index = index + 1
    }
    
    for ptr1 < len(nums1) {
        merged[index] = nums1[ptr1]
        ptr1 = ptr1 + 1
        index = index + 1
    }
    
    for ptr2 < len(nums2) {
        merged[index] = nums2[ptr2]
        ptr2 = ptr2 + 1
        index = index + 1
    }

    return getMedian(merged)
}
